### DISCUSSION TITLE
Examples of working with localStorage?

### DISCUSSION THREAD

**Discussion by tanmdoan** [johnlindquist/kit#629] [Q&A] [answered]
Hello! I'm looking to create a helper which involves setting a localStorage item for chrome browser instance, I tried searching around without any luck, is this something that's even possible? If so would appreciate some references! 

Added context:

Manual steps we have to do constantly: 

1. open browser, go to a link
2. grab the value via local storage
3. go back to localhost and set the value manually in localStorage
4. reload the browser

Thankfully we have some bookmarks that execute javascript, but it's still 3-4 steps that I would love to be able to run a script to do. 

Thanks!

**Comment by johnlindquist** [comment, marked as answer]
Hi @tanmdoan , great question:

Here's the basics of executing JS in Chrome from Script Kit:

## Enable JS From Apple Events
![CleanShot 2021-11-30 at 18 21 33](https://user-images.githubusercontent.com/36073/144154886-fa3bb56d-4020-48fe-a03c-e63abb9375fa.png)


## Pass a String of JS to AppleScript

[Install modify-local-storage](https://scriptkit.com/api/new?name=modify-local-storage&url=https://gist.githubusercontent.com/johnlindquist/dbb32f90437bf7ee376a39fa7b8d28a9/raw/d4fc575f2065b90d1f9ddaa8a1e3e9dfb6f74487/modify-local-storage.js")

```js
// Name: Modify Local Storage

let js = `

localStorage.setItem('kit', 'is awesome!!!');

`

let value = await applescript(`
tell application "Google Chrome" to tell window 1
	set str to execute active tab javascript "${js}"
	return str
end tell
`)
```

That should get you on the right track.

Would you like me to show you how to do the other steps as well, or would you prefer to figure it out yourself?

**Reply by tanmdoan** [reply]
Amazing, thank you so much for the prompt response. Let me see if I can figure it out, I'll let you know if I get stuck! Appreciate it!

**Reply by tanmdoan** [reply]
@johnlindquist I'm really close, I seem to be running into a similar issue as here (https://github.com/johnlindquist/kit/discussions/64) -- I am getting a 400 error, I'm not tracking what you mean with your answer. Here's my code: 

```javascript
import '@johnlindquist/kit'
let TMC_API_TOKEN = await env('TMC_API_TOKEN')

// Name: tmc-login

let config = {
  headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
}

let response = await post(
  'https://console-stg.cloud.vmware.com/csp/gateway/am/api/auth/api-tokens/authorize',
  {
    body: {
      redacted: TMC_API_TOKEN,
    },
  },
  config
)

let js = `localStorage.setItem('redacted', "${response.body.redacted}")`

let value = await applescript(`
tell application "Google Chrome" to tell window 1
	set str to execute active tab javascript "${js}"
	return str
end tell
`)

exec(`open "http://localhost:3001"`)

```

**Reply by johnlindquist** [reply]
`post` uses Axios's post. So examples like these should work: https://stackoverflow.com/questions/44617825/passing-headers-with-axios-post-request


I'm guessing that 

```js
let response = await post(
  'https://console-stg.cloud.vmware.com/csp/gateway/am/api/auth/api-tokens/authorize',
  {
    body: {
      redacted: TMC_API_TOKEN,
    },
  },
  config
)
```

Should remove the "body"

```js
let response = await post(
  'https://console-stg.cloud.vmware.com/csp/gateway/am/api/auth/api-tokens/authorize',
  {
      redacted: TMC_API_TOKEN,
  },
  config
)
```

**Reply by johnlindquist** [reply]
node-fetch is also available on the global API if you're more comfortable with that. Just use `fetch`

**Reply by tanmdoan** [reply]
ah, yea let me give that a shot!

**Comment by tanmdoan** [comment]
@johnlindquist just wanted to give you an update -- got the script working exactly like how we wanted. Currently ramping up my team with it and have been getting very positive responses so far. Thank you so much for your help, and creating such an amazing tool. Next we'll be exploring a more seamless way to share scripts which I saw in another thread that you will be releasing a demo/info about shortly. Cheers!

**Reply by tanmdoan** [reply]
another update, currently exploring the `Managing kenvs` option in the app.

**Reply by johnlindquist** [reply]
Would love to hear about your experience. I need to document those workflows still and see how people use them.

**Reply by tanmdoan** [reply]
So, our first stab at it before we discovered the managing kenvs options, was to just check in the whole `.kenv` folder but soon we found that to not be ideal mainly for obvious reasons like installation/configuring created overlapping folders/files. There were lots of manual steps including having to create necessary `.gitignore`d files or running `npm i`. 

After digging into the `Manage kenvs` feature, we ended up just checking into git the appropriate `kenvs/` however the clone via remote feature only works for github and we actually use gitlab at VMware so that threw some errors for us. Not wanting to dig into that initially, we took a different approach by cloning the `kenv` locally then linked it via the path. This worked ok but seemed a bit complicated still, we thought we could do better.

Finally, after exploring what went wrong in the clone script/logs, we chose to modify the clone script as part of the instructions before linking to our remote `kenv`. Not ideal, but this seemed like a good short term compromise, and we do realize this might be a great `OSS` contribution opportunity for the future. Not sure what the ideal DX for that would look like but maybe taking an exact full `git` url, or a flag to which version control service like `tanmdoan/my-kenv --gitlab` or something would be nice. 

Anyway, as I've mentioned before, super appreciate the work that has gone into ScriptKit, and specifically small but important details like thoughtful debugging/logging. Looking forward to see some demos/examples in the future and iterating on the process!

Kindly,
Tan

**Reply by johnlindquist** [reply]
Hmm, the `kenv-clone` script just runs `git clone`, so it "should" support gitlab:

https://github.com/johnlindquist/kit/blob/beta/src/cli/kenv-clone.ts#L69

Do you have a demo kenv on gitlab I could test?

Could you share what changes to made to make it work? Thanks!

**Reply by tanmdoan** [reply]
I wonder if it's because I didn't use the `https` url -- the gitlab repo is private unfortunately and namespace to our org but here's the error in the logs:

```
[2021-12-02 10:00:20.475] [info]  $ git clone $'git@gitlab.eng.vmware.com:olympus/tmc-script-kit.git' /Users/tandoan/.kenv/kenvs/testing
```

**Reply by tanmdoan** [reply]
yea that makes sense actually, `git@gitlab.eng.vmware.com:olympus/tmc-script-kit.git` splits into 2 😆 -- user error

**Reply by johnlindquist** [reply]
Ooooh, whoops 🤦‍♂️. I'll have to add some smarter regex! Sorry about that.

**Reply by tanmdoan** [reply]
haha all good! Shoulda been better detectives haha..
